# Titanic Dataset

In [138]:
# import libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [140]:
train_df = pd.read_csv("train.csv")
train_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [142]:
test_df = pd.read_csv("test.csv")
test_df.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

In [144]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [146]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  418 non-null    int64  
 1   Pclass       418 non-null    int64  
 2   Name         418 non-null    object 
 3   Sex          418 non-null    object 
 4   Age          332 non-null    float64
 5   SibSp        418 non-null    int64  
 6   Parch        418 non-null    int64  
 7   Ticket       418 non-null    object 
 8   Fare         417 non-null    float64
 9   Cabin        91 non-null     object 
 10  Embarked     418 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 36.0+ KB


In [148]:
train_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

- Survived is a binary feature with 0 or 1
- On average ~38% of passengers survived
- Average age was ~30 years old
- Very few elderly people onboard

In [150]:
train_df.describe(include=["O"])

Name   Sex    Ticket        Cabin Embarked
count                       891   891       891          204      889
unique                      891     2       681          147        3
top     Hansen, Mr. Claus Peter  male  CA. 2343  C23 C25 C27        S
freq                          1   577         7            4      644

- Everyone had a unique name
- The majority of passengers were male
- Duplication or sharing of cabins
- 3 locations where passengers embarked with S being the most frequently occuring

In [152]:
train_df[["Sex", "Survived"]].groupby(['Sex'], as_index=False).mean()

Sex  Survived
0  female  0.742038
1    male  0.188908

In [154]:
train_df[['Pclass', 'Survived']].groupby(['Pclass'], as_index=False).mean()

Pclass  Survived
0       1  0.629630
1       2  0.472826
2       3  0.242363

So it seems that both the sex of a passenger and their class affected their chances of survival with females surviving almost 75% of the time compared to 19% for males and 1st class passengers having a survival rate of around 63% compared to 47% and 24% for 2nd and 3rd class respectfully.

In [156]:
train_df[["SibSp", "Survived"]].groupby(['SibSp'], as_index=False).mean()

SibSp  Survived
0      0  0.345395
1      1  0.535885
2      2  0.464286
3      3  0.250000
4      4  0.166667
5      5  0.000000
6      8  0.000000

In [158]:
train_df[["Parch", "Survived"]].groupby(['Parch'], as_index=False).mean()

Parch  Survived
0      0  0.343658
1      1  0.550847
2      2  0.500000
3      3  0.600000
4      4  0.000000
5      5  0.200000
6      6  0.000000

On the other hand, it seems like the number of family members aboard a passenger had did not significantly affect their chances of survival an so may not be a useful feature for model development.

Next I'm going to try find if there is a correlation between age group and survival.
For the age groups in years I will be basing it on these ranges:

| 0-14  	| Child       	|
|-------	|:-------------:|
| 15-24 	| Youth       	|
| 25-44 	| Young Adult 	|
| 45-64 	| Adult       	|
| 65+   	| Elder       	|

In [160]:
bins= [0, 14, 24, 44, 64, 150]
labels = ["Child","Youth","Young Adult","Adult","Elder"]
train_df["Age Group"] = pd.cut(train_df["Age"], bins=bins, labels=labels, right=True)
train_df.head()

Survived  Pclass     Sex   Age     Fare    Age Group
0         0       3    male  22.0   7.2500        Youth
1         1       1  female  38.0  71.2833  Young Adult
2         1       3  female  26.0   7.9250  Young Adult
3         1       1  female  35.0  53.1000  Young Adult
4         0       3    male  35.0   8.0500  Young Adult

In [162]:
train_df.groupby("Age Group").mean()[["Survived"]]

Survived
Age Group            
Child        0.584416
Youth        0.365000
Young Adult  0.400621
Adult        0.403846
Elder        0.090909

So it seems like children have a much higher survival rate than other age group with around 58.4% surviving which reinforces the children and woman leave the boat first. Another stat that stands out are the elders who only had around 9.1% of them surviving. This is probably due to their bodies being less than optimal for surviving the sinking. Age could be a feature for the predictive model.

In [164]:
grid = sns.FacetGrid(train_df, row='Embarked', size=4, aspect=1.5)
grid.map(sns.lineplot, 'Pclass', 'Survived', 'Sex')
grid.add_legend()

KeyError: 'Embarked'

- Females had a much higher survival rate
- C port is the exception where had a higher survival rate, may just be a correlation between port and Pclass

I might not use the Embarked feature for the model

In [159]:
train_df.drop(["PassengerId", "Name", "SibSp", "Parch", "Ticket", "Cabin", "Embarked", "Age Group"], axis=1, inplace=True)
train_df.head()

Survived  Pclass     Sex   Age     Fare
0         0       3    male  22.0   7.2500
1         1       1  female  38.0  71.2833
2         1       3  female  26.0   7.9250
3         1       1  female  35.0  53.1000
4         0       3    male  35.0   8.0500

In [161]:
test_df.drop(["PassengerId", "Name", "SibSp", "Parch", "Ticket", "Cabin", "Embarked"], axis=1, inplace=True)
test_df.head()

Pclass     Sex   Age     Fare
0       3    male  34.5   7.8292
1       3  female  47.0   7.0000
2       2    male  62.0   9.6875
3       3    male  27.0   8.6625
4       3  female  22.0  12.2875

In [163]:
train_df["Sex"] = train_df["Sex"].replace({"male": 0, "female": 1})
train_df.head()

Survived  Pclass  Sex   Age     Fare    Age Group
0         0       3    0  22.0   7.2500        Youth
1         1       1    1  38.0  71.2833  Young Adult
2         1       3    1  26.0   7.9250  Young Adult
3         1       1    1  35.0  53.1000  Young Adult
4         0       3    0  35.0   8.0500  Young Adult

In [165]:
test_df["Sex"] = test_df["Sex"].replace({"male": 0, "female": 1})
test_df.head()

Pclass  Sex   Age     Fare
0       3    0  34.5   7.8292
1       3    1  47.0   7.0000
2       2    0  62.0   9.6875
3       3    0  27.0   8.6625
4       3    1  22.0  12.2875

In [166]:
train_df["Age"].fillna(train_df["Age"].median())

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888    28.0
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [167]:
train_df.loc[train_df["Age"] <= 14, "Age"] = 0
train_df.loc[(14 < train_df["Age"]) & (train_df["Age"] <= 24), "Age"] = 1
train_df.loc[(24 < train_df["Age"]) & (train_df["Age"]<= 45), "Age"] = 2
train_df.loc[(45 < train_df["Age"]) & (train_df["Age"]<= 64), "Age"] = 3
train_df.loc[train_df["Age"] > 64, "Age"] = 4

In [168]:
test_df.loc[test_df["Age"] <= 14, "Age"] = 0
test_df.loc[(14 < test_df["Age"]) & (test_df["Age"] <= 24), "Age"] = 1
test_df.loc[(24 < test_df["Age"]) & (test_df["Age"]<= 45), "Age"] = 2
test_df.loc[(45 < test_df["Age"]) & (test_df["Age"]<= 64), "Age"] = 3
test_df.loc[test_df["Age"] > 64, "Age"] = 4

In [169]:
train_df.head()

Survived  Pclass  Sex  Age     Fare    Age Group
0         0       3    0  1.0   7.2500        Youth
1         1       1    1  2.0  71.2833  Young Adult
2         1       3    1  2.0   7.9250  Young Adult
3         1       1    1  2.0  53.1000  Young Adult
4         0       3    0  2.0   8.0500  Young Adult

In [170]:
test_df.head()

Pclass  Sex  Age     Fare
0       3    0  2.0   7.8292
1       3    1  3.0   7.0000
2       2    0  3.0   9.6875
3       3    0  2.0   8.6625
4       3    1  1.0  12.2875